apps/prairielearn/src/lib/authzDataPublishing.ts

```diff
@@ -0,0 +1,205 @@
+ * This function is only used by authzData.ts
+ *
+ * @param courseInstance - The course instance to check access for.
+ * @param authzData - The authorization data of the user.
+ * @param reqDate - The date of the request.
+ */
+export async function calculateModernCourseInstanceStudentAccess(
+  courseInstance: CourseInstance,
+  authzData: RawPageAuthzData,
+  reqDate: Date,
+) {
+  const enrollment = await selectOptionalEnrollmentByUserId({
+    userId: authzData.user.user_id,
+    requestedRole: 'Student',
+    authzData,
+    courseInstance,
+  });
+
+  // We only consider non-legacy publishing.
+  if (!courseInstance.modern_publishing) {
+    throw new Error('Course instance is not using modern publishing');
+  }
+
+  // Not published at all.
+  if (courseInstance.publishing_start_date == null) {
+    return { has_student_access: false, has_student_access_with_enrollment: false };
+  }
+  // End date is always set alongside start date
+  assert(courseInstance.publishing_end_date != null);
+
+  // Before the start date, we definitely don't have access.
+  if (reqDate < courseInstance.publishing_start_date) {
+    return { has_student_access: false, has_student_access_with_enrollment: false };
+  }
+
+  //If we are before the end date and after the start date, we definitely have access.
+  if (reqDate < courseInstance.publishing_end_date) {
+    return { has_student_access: true, has_student_access_with_enrollment: enrollment != null };
+  }
+
+  // We are after the end date. We might have access if we have an extension.
+  const publishingExtensions =
+    enrollment != null ? await selectPublishingExtensionsByEnrollmentId(enrollment.id) : [];
+
+  // There are no extensions. We don't have access.
+  if (publishingExtensions.length === 0) {
+    return { has_student_access: false, has_student_access_with_enrollment: false };
+  }
+
+  // Consider the latest enabled date for the enrollment.
+  const allDates = publishingExtensions
+    .map((extension) => extension.end_date)
+    .sort((a, b) => {
+      return b.getTime() - a.getTime();
+    });
+  const latestDate = allDates[0];
+
+  // If we are before the latest date, we have access.
+  return {
+    has_student_access: reqDate < latestDate,
+    has_student_access_with_enrollment: reqDate < latestDate && enrollment != null,
+  };
+}
+
+/**
+ * Builds the authorization data for a user on a page.
+ *
+ * @param params
+ * @param params.authn_user - The authenticated user.
+ * @param params.course_id - The ID of the course. If not provided,
+ * but course_instance_id is provided, the function will use the course_id from the course instance.
+ * @param params.course_instance_id - The ID of the course instance.
+ * @param params.is_administrator - Whether the user is an administrator.
+ * @param params.ip - The IP address of the request.
+ * @param params.req_date - The date of the request.
+ * @param params.req_mode - The mode of the request.
+ */
+export async function buildAuthzData({
```

Comment:
I don't understand why this function as a whole was moved into `authzDataPublishing.ts` - can you explain? This felt much more at home in `authzData.ts`. Moving this also makes this harder to review.